src/Modules/ModulesGraded.jl

```diff
@@ -1,5 +1,696 @@
+  T2 = codomain(f)
+  if !is_graded(T1) || !is_graded(T2)
+    error("Both domain and codomain must be graded.")
+  end
+  if iszero(T1)
+    R = base_ring(T1)
+    G = grading_group(R)
+    return G[0]
+  end
+  domain_degrees = degrees_of_generators(T1)
+  df = nothing
+  for i in 1:length(domain_degrees)
+    image_vector = f(T1[i])
+    if isempty(coordinates(image_vector))
+      continue
+    end
+    current_df = degree(image_vector) - domain_degrees[i]
+    if df === nothing
+      df = current_df
+    elseif df != current_df
+      error("The homomorphism is not graded.")
+    end
+  end
+  if df === nothing
+    R = base_ring(T1)
+    G = grading_group(R)
+    return G[0]
+  end
+  return df
+end
+
+function is_graded(f::SubQuoHom)
+  return isdefined(f, :d)
+end
+
+function grading_group(f::SubQuoHom)
+  return grading_group(base_ring(domain(f)))
+end
+
+function is_homogeneous(f::SubQuoHom)
+  A = grading_group(f)
+  return isdefined(f, :d) && degree(f)==A[0]
+end
+
+
+###############################################################################
+# Graded free resolutions
+###############################################################################
+
+function is_graded(resolution::FreeResolution{T}) where T
+  C = resolution.C
+ return all(is_graded(C[i]) for i in reverse(Hecke.range(C))) && all(is_graded(map(C, i)) for i in reverse(Hecke.map_range(C)))
+end
+
+###############################################################################
+# Betti table
+###############################################################################
+
+function betti_table(F::FreeResolution; project::Union{GrpAbFinGenElem, Nothing} = nothing, reverse_direction::Bool=false)
+  generator_count = Dict{Tuple{Int, Any}, Int}()
+  C = F.C
+  rng = Hecke.map_range(C)
+  n = first(rng)
+  for i in 0:n
+      module_degrees = F[i].d
+      module_degrees === nothing && error("One of the modules in the graded free resolution is not graded.")
+      for degree in module_degrees
+          idx = (i, degree)
+          generator_count[idx] = get(generator_count, idx, 0) + 1
+      end
+  end
+  return BettiTable(generator_count, project = project, reverse_direction = reverse_direction)
+end
+
+function betti(b::FreeResolution; reverse_direction::Bool = false)
+	return betti_table(b, project = nothing, reverse_direction = reverse_direction)
+end
+
+function dict(b::BettiTable)
+  return b.B
+end
+
+function reverse_direction!(b::BettiTable)
+  if b.reverse_direction
+    b.reverse_direction = false
+  else
+    b.reverse_direction = true
+  end
+  return
+end
+
+function induce_shift(B::Dict{Tuple{Int, Any}, Int})
+  A = parent(first(keys(B))[2])
+  new_B = Dict{Tuple{Int, Any}, Int}()
+  for ((i, key), value) in B
+      new_key = (i, key-i*A[1])
+      new_B[new_key] = value
+  end
+  return new_B
+end
+
+function Base.show(io::IO, b::BettiTable)
+  T = induce_shift(b.B)
+  x = collect(keys(T))
+  isempty(x) && (println("Empty table"); return)
```

Comment:
```suggestion
  isempty(x) && (println(io, "Empty table"); return)
```
Stylistically, I'd really prefer to spell it out:
```suggestion
  if isempty(x)
    println(io, "Empty table")
    return
  end
```
Perhaps it should also say "Empty Betti table"?

Reply:
Improved the Betti table code, further improvements for the case of using a projection will come.